File: ios/TVRO/UpdatesManager.m

#import "UpdatesManager.h"
#import "UpdatesManagerDelegate.h"

@interface UpdatesManager ()

@end

@implementation UpdatesManager

#pragma mark - UpdatesManager methods

- (id)initWithDelegate:(id<UpdatesManagerDelegate>)_delegate {
	self = [super init];
	
	delegate = _delegate;
	
	NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
	if (![defaults stringForKey:@"satlibrary-device-version"]) [defaults setValue:@"" forKey:@"satlibrary-device-version"];
	if (![defaults stringForKey:@"tv1-device-version"]) [defaults setValue:@"" forKey:@"tv1-device-version"];
	if (![defaults stringForKey:@"tv3-device-version"]) [defaults setValue:@"" forKey:@"tv3-device-version"];
	if (![defaults stringForKey:@"tv5-device-version"]) [defaults setValue:@"" forKey:@"tv5-device-version"];
	if (![defaults stringForKey:@"tv6-device-version"]) [defaults setValue:@"" forKey:@"tv6-device-version"];
	if (![defaults stringForKey:@"rv1-device-version"]) [defaults setValue:@"" forKey:@"rv1-device-version"];
	
	fileData = [[NSMutableData alloc] init];
	alertView = [[UIAlertView alloc] initWithTitle:@"Downloading" message:@"Please wait!" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:nil];
	
	return self;
}

- (void)startDownloadForUpdateType:(NSString *)_updateType portalVersion:(NSString *)_portalVersion portalUrl:(NSURL *)portalUrl {
	if (connection) {
		[connection cancel];
		connection = nil;
	}
	
	[fileData setLength:0];
	
  //	reinforce lowercase key names
  //	web code might send us 'TV5' instead of 'tv5'
	updateType = [_updateType lowercaseString];
	portalVersion = _portalVersion;
	
	connection = [[NSURLConnection alloc] initWithRequest:[NSURLRequest requestWithURL:portalUrl] delegate:self];
	[connection start];
	
	[alertView show];
}

- (void)cancelDownload {
	if (connection) {
		[connection cancel];
		connection = nil;
	}
	
	[fileData setLength:0];
	
	if (delegate)
		if ([delegate respondsToSelector:@selector(updatesManager:downloadCancelledForUpdateType:)])
			[delegate updatesManager:self downloadCancelledForUpdateType:updateType];
}

- (void)startUploadForUpdateType:(NSString *)_updateType uploadUrl:(NSURL *)uploadUrl {
	
	if (connection) {
		[connection cancel];
		connection = nil;
	}
	
	[fileData setLength:0];
	
	updateType = [_updateType lowercaseString];
	
	NSString* filePath = [self filePathForUpdateType:updateType];
	NSString* fileName = [filePath lastPathComponent];
	[fileData appendData:[NSData dataWithContentsOfFile:filePath]];

	NSLog(@"filePath: %@", filePath);
	NSLog(@"fileName: %@", fileName);
	
	NSString* stringBoundary = @"0xKHTMLBoundary";
  NSMutableDictionary* headers = [[NSMutableDictionary alloc] init];
  [headers setValue:@"no-cache" forKey:@"Cache-Control"];
  [headers setValue:@"no-cache" forKey:@"Pragma"];
  [headers setValue:[NSString stringWithFormat:@"multipart/form-data; boundary=%@", stringBoundary] forKey:@"Content-Type"];
  
  NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:uploadUrl cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60.0];
  [request setHTTPMethod:@"POST"];
  [request setAllHTTPHeaderFields:headers];
  
  NSMutableData* postData = [NSMutableData dataWithCapacity:[fileData length] + 512];
  [postData appendData:[[NSString stringWithFormat:@"--%@\r\n", stringBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
  [postData appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"fileToUpload\"; filename=\"%@\"\r\n\r\n", fileName] dataUsingEncoding:NSUTF8StringEncoding]];
  [postData appendData:fileData];
  [postData appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n", stringBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
  [request setHTTPBody:postData];
	
	connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
	[connection start];
}

- (void)cancelUpload {
	if (connection) {
		[connection cancel];
		connection = nil;
	}
	
	if (delegate)
		if ([delegate respondsToSelector:@selector(updatesManager:uploadCancelledForUpdateType:)])
			[delegate updatesManager:self uploadCancelledForUpdateType:updateType];
}

- (NSString*)deviceVersionForUpdateType:(NSString *)_updateType {
  NSString* key = [NSString stringWithFormat:@"%@-device-version", [_updateType lowercaseString]];
	return [[NSUserDefaults standardUserDefaults] stringForKey:key];
}

- (NSString*)filePathForUpdateType:(NSString *)_updateType {
	return [[NSUserDefaults standardUserDefaults] stringForKey:[NSString stringWithFormat:@"%@-file-path", [_updateType lowercaseString]]];
}

#pragma mark - NSURLConnectionDelegate protocol methods

- (void)connection:(NSURLConnection *)_connection didReceiveResponse:(NSURLResponse *)response {
	NSLog(@"connection:%@ didReceiveResponse:%@", _connection, response);
}

- (void)connection:(NSURLConnection *)_connection didReceiveData:(NSData *)data {
	NSLog(@"connection:%@ didReceiveData:%d", _connection, [data length]);
  [fileData appendData:data];
}

- (void)connection:(NSURLConnection *)_connection didFailWithError:(NSError *)error {
	NSLog(@"connection:%@ didFailWithError:%@", _connection, error);
}

- (void)connectionDidFinishLoading:(NSURLConnection *)_connection {
	NSLog(@"connectionDidFinishLoading:%@", _connection);
	NSString* fileName = connection.currentRequest.URL.lastPathComponent;
	NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
	NSString *documentsDirectory = [paths objectAtIndex:0];
	NSString *filePath = [NSString stringWithFormat:@"%@/%@/%@", documentsDirectory, updateType, fileName];
		
	[fileData writeToFile:filePath atomically:YES];
  [alertView dismissWithClickedButtonIndex:0 animated:YES];
	
	NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
	[defaults setValue:portalVersion forKey:[NSString stringWithFormat:@"%@-device-version", updateType]];
	[defaults setValue:filePath forKey:[NSString stringWithFormat:@"%@-file-path", updateType]];
	[defaults synchronize];
	
	if (delegate)
		if ([delegate respondsToSelector:@selector(updatesManager:downloadCompletedForUpdateType:)])
			[delegate updatesManager:self downloadCompletedForUpdateType:updateType];
}

#pragma mark - UIAlertViewDelegate protocol methods

- (void)alertView:(UIAlertView *)_alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
	
}

@end

File: android/src/com/kvh/kvhandroid/utils/UpdatesManager.java
package com.kvh.kvhandroid.utils;

import java.io.File;

import android.app.Activity;
import android.app.DownloadManager;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Environment;
import android.util.Log;

public class UpdatesManager {
	private String TAG = "KVHANDROID - UpdatesManager";
	
	private Activity activity;
	private UpdatesManagerCallback updatesManagerCallback;
	private DownloadManager downloadManager;
	
	private String updateType;
	private String portalVersion;
	private long download_id;
	
	private ProgressDialog downloadProgressDialog;
	
	public UpdatesManager(Activity act, UpdatesManagerCallback callback) {
		updatesManagerCallback = callback;
		this.activity = act;
		
		downloadManager = (DownloadManager)activity.getSystemService(activity.DOWNLOAD_SERVICE);

	}
	
	public String deviceVersionForUpdateType(String updateType) {
		String key =  updateType.toLowerCase() + "-device-version";
		
		SharedPreferences settings = activity.getSharedPreferences(Constants.PREFS_NAME, 0);
		return settings.getString(key, "");
	}
	
	public void startDownload(String _updateType, String _portalVersion, String portalURL) {
		String tempPortalURL = portalURL;
		updateType = _updateType.toLowerCase();
		portalVersion = _portalVersion.toLowerCase();
		
		//setup the download rpogress dialog
		downloadProgressDialog = new ProgressDialog(activity);
		downloadProgressDialog.setTitle("Downloading Update ...");
		downloadProgressDialog.setMessage("Download in progress ...");
		downloadProgressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
		downloadProgressDialog.show();
		
		//fix the download url
		if(tempPortalURL.charAt(tempPortalURL.length() - 1) == '/')
			tempPortalURL = tempPortalURL.substring(0, tempPortalURL.length() - 1);
		
		Uri download_uri = Uri.parse(tempPortalURL);
		
		Log.i(TAG, "Download Uri: " + download_uri.getHost() + " " + download_uri.getPath() + " " + download_uri.getPathSegments().get(download_uri.getPathSegments().size() - 1));
		
		//create the subfolders
		File direct = new File(Environment.getExternalStorageDirectory() + "/kvh_files");
        if (!direct.exists()) {
            direct.mkdirs();
        }
		
		DownloadManager.Request request = new DownloadManager.Request(download_uri);
		//set the download path
		request.setDestinationInExternalPublicDir("/kvh_files", download_uri.getPathSegments().get(download_uri.getPathSegments().size() - 1));
		//start downloading
		download_id = downloadManager.enqueue(request);
		
		//Save the download id
		SharedPreferences settings = activity.getSharedPreferences(Constants.PREFS_NAME, 0);
		SharedPreferences.Editor editor = settings.edit();
		editor.putLong("downloadid", download_id);
		// Commit the edits!
		editor.commit();
		
		//to get the download progress
		new Thread(new Runnable() {
            @Override
            public void run() {

                boolean downloading = true;

                while (downloading) {

                    DownloadManager.Query q = new DownloadManager.Query();
                    q.setFilterById(download_id);

                    android.database.Cursor cursor = downloadManager.query(q);
                    cursor.moveToFirst();
                    final int bytes_downloaded = cursor.getInt(cursor
                            .getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
                    final int bytes_total = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));

                    if (cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)) == DownloadManager.STATUS_SUCCESSFUL) {
                        downloading = false;
                    }
                    
                    Log.d(TAG, "Downloaded: " + bytes_downloaded + " / " + bytes_total);

//                    final int dl_progress = (int) ((bytes_downloaded * 100l) / bytes_total);
                    
                    activity.runOnUiThread(new Runnable() {

                        @Override
                        public void run() {
                        	if(bytes_total == -1) {
                        		downloadProgressDialog.setProgress(0);
                        		downloadProgressDialog.setMax(1);
                        	} else {
                        		//if(!downloadProgressDialog.isShowing())
                        		downloadProgressDialog.setProgress(bytes_downloaded);
                        		downloadProgressDialog.setMax(bytes_total);
                        	}
                        	
                        }
                    });
                    cursor.close();
                }

            }
        }).start();
	}
	
	public void startUpload(String _updateType, String uploadUrl) {
		/*
		if (connection) {
			[connection cancel];
			connection = nil;
		}
		
		[fileData setLength:0];
		
		updateType = [_updateType lowercaseString];
		
		NSString* filePath = [self filePathForUpdateType:updateType];
		NSString* fileName = [filePath lastPathComponent];
		[fileData appendData:[NSData dataWithContentsOfFile:filePath]];
		
		NSString* stringBoundary = @"0xKHTMLBoundary";
	  NSMutableDictionary* headers = [[NSMutableDictionary alloc] init];
	  [headers setValue:@"no-cache" forKey:@"Cache-Control"];
	  [headers setValue:@"no-cache" forKey:@"Pragma"];
	  [headers setValue:[NSString stringWithFormat:@"multipart/form-data; boundary=%@", stringBoundary] forKey:@"Content-Type"];
	  
	  NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:uploadUrl cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60.0];
	  [request setHTTPMethod:@"POST"];
	  [request setAllHTTPHeaderFields:headers];
	  
	  NSMutableData* postData = [NSMutableData dataWithCapacity:[fileData length] + 512];
	  [postData appendData:[[NSString stringWithFormat:@"--%@\r\n", stringBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
	  [postData appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"fileToUpload\"; filename=\"%@\"\r\n\r\n", fileName] dataUsingEncoding:NSUTF8StringEncoding]];
	  [postData appendData:fileData];
	  [postData appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n", stringBoundary] dataUsingEncoding:NSUTF8StringEncoding]];
	  [request setHTTPBody:postData];
	  
	  [uploadAlertView show];
	  
	  uploading = true;
		connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
		[connection start];
		*/
	}
	
	public void registerDownloadManager() {   
		IntentFilter intentFilter = new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE);
		activity.registerReceiver(downloadReceiver, intentFilter);
	}
	
	public void unregisterDownloadManager() {
		activity.unregisterReceiver(downloadReceiver);
	}
	  
	 private BroadcastReceiver downloadReceiver = new BroadcastReceiver() {

		@Override
		public void onReceive(Context context, Intent intent) {
			   
			DownloadManager.Query query = new DownloadManager.Query();
			SharedPreferences settings = context.getSharedPreferences(Constants.PREFS_NAME, 0);
			query.setFilterById(settings.getLong("downloadid", 0));
			android.database.Cursor cursor = downloadManager.query(query);
			     
			if(cursor.moveToFirst()) {
				int columnIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
			    int status = cursor.getInt(columnIndex);
			    int columnReason = cursor.getColumnIndex(DownloadManager.COLUMN_REASON);
			    int reason = cursor.getInt(columnReason);
			      
			    if(status == DownloadManager.STATUS_SUCCESSFUL) {  	
			    	String fileLocationString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
			    	
			    	//save information in preferences
			    	SharedPreferences.Editor editor = settings.edit();
		    		editor.putString(updateType + "-device-version", portalVersion);
		    		editor.putString(updateType + "-file-path", fileLocationString);
		    		editor.commit();
			       
		    		downloadProgressDialog.dismiss();
		    		updatesManagerCallback.downloadCompleted();
		    		cursor.close();
			    } else if(status == DownloadManager.STATUS_FAILED) {
			    	android.widget.Toast.makeText(context, "FAILED!\n" + "reason of " + reason, android.widget.Toast.LENGTH_LONG).show();
			    } else if(status == DownloadManager.STATUS_PAUSED) {
			    	android.widget.Toast.makeText(context, "PAUSED!\n" + "reason of " + reason, android.widget.Toast.LENGTH_LONG).show();
			    } else if(status == DownloadManager.STATUS_PENDING) {
			    	android.widget.Toast.makeText(context, "PENDING!", android.widget.Toast.LENGTH_LONG).show();
			    }else if(status == DownloadManager.STATUS_RUNNING){
//			    	android.widget.Toast.makeText(context, "RUNNING!", android.widget.Toast.LENGTH_LONG).show();
			    	int bytes_downloaded = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
                    int bytes_total = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
			    	Log.i(TAG, "Download Manager Running: " + bytes_downloaded + " / " + bytes_total);
			    }
			}
		}
	    
	 };
}
